[android] fix: runtimeVersion not visible in dev-client #34090

chrfalch opened the pull request:
# Why 

#34090 reports that the field `runtimeVersion` in the `expo-dev-client` wasn't displaying correctly under the old architecture on Android - it is fine and visible on the new architecture on both Android and iOS.

This was eventually confirmed after some back and forth I ended up with the following reasoning:

The field `runtimeVersion` comes from the `expo-updates` package and is read in `DevLauncherInternalModule.getUpdatesConfig()` method using the following statement:

```kt
private fun getUpdatesConfig(): WritableMap {
  // ...
  val runtimeVersion = controller.updatesInterface?.runtimeVersion
  // ...
}
```

The problem was that the dependency chain required a specific order of initialisation to ensure that both `controller` and `controller.updatesInterface` was correctly set before this would work.

The `controller` field is set in the `UpdatesController.initialiseWithoutStarting` method:

```kt
@JvmStatic fun initializeWithoutStarting(context: Context) {
  // ...
  if (BuildConfig.USE_DEV_CLIENT) {
    val devLauncherController = initializeAsDevLauncherWithoutStarting(context)
    singletonInstance = devLauncherController
    UpdatesControllerRegistry.controller = WeakReference(devLauncherController) // <----- HERE
  } else {
    singletonInstance = DisabledUpdatesController(context, null)
  }
  // ...
}
```

This code is called after we initialise the `DevLauncherController` and initialize the instance's `updatesInterface` member:

```kt
fun createApplicationLifecycleListeners(context: Context?): List<ApplicationLifecycleListener> =
  listOf(
    object : ApplicationLifecycleListener {
      override fun onCreate(application: Application?) {
        check(application is ReactApplication)
        DevLauncherController.initialize(application)
        UpdatesControllerRegistry.controller?.get()?.let {
          DevLauncherController.instance.updatesInterface = it  // <----- HERE
          it.updatesInterfaceCallbacks = WeakReference(DevLauncherController.instance)
        }
      }
    }
  )
```

For this to work the instance of the `DevLauncherController` has to be set - which is done as a sideeffect in `DevLauncherController.initialize(application)`.

*The problem is that this side effect was not called on the old architecture.*

The side effect is called from within the `ReactHostWrapper.kt` when calling the `reactHostProvider()` function argument on initialisation. When using the old architecture this was not done as a solution to a problem with multiple reloads that was addressed in #32532.

# How

This commit proposes a solution to this where the `DevLauncherController` making sure the necessary side effects is run at the correct time:

```kt
@JvmStatic
fun initialize(reactApplication: ReactApplication, additionalPackages: List<ReactPackage>? = null, launcherClass: Class<*>? = null) {
  // ...
  val reactHost = reactApplication.reactHost
  initialize(reactApplication as Context, ReactHostWrapper(reactApplication.reactNativeHost) { reactHost })
  // ...
}
```

Here we're calling the `reactHost` property *only once* and then pass a new lambda to the ReactHostWrapper so it can initialize without having to consider these side effects.

Closes #ENG-14823

# Test Plan

- We should be able to see the `runtimeVersion` field in the dev-client as noted in the original issue when running with the old architecture.
- We should be able to reload on the old architecture without the app reloading multiple times (#32532)

# Checklist

- [X] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

chrfalch committed:
[android] fix: runtimeVersion not visible in dev-client

#34090 reports that the field `runtimeVersion` in the `expo-dev-client` wasn't displaying correctly under the old architecture on Android - it is fine and visible on the new architecture on both Android and iOS.

This was eventually confirmed after some back and forth I ended up with the following reasoning:

The field `runtimeVersion` comes from the `expo-updates` package and is read in `DevLauncherInternalModule.getUpdatesConfig()` method using the following statement:

```kt
private fun getUpdatesConfig(): WritableMap {
    ...
    val runtimeVersion = controller.updatesInterface?.runtimeVersion
    ...
}
```

The problem was that the dependency chain required a specific order of initialisation to ensure that both `controller` and `controller.updatesInterface` was correctly set before this would work.

The `controller` field is set in the `UpdatesController.initialiseWithoutStarting` method:

```kt
@JvmStatic fun initializeWithoutStarting(context: Context) {
      ...
        if (BuildConfig.USE_DEV_CLIENT) {
          val devLauncherController = initializeAsDevLauncherWithoutStarting(context)
          singletonInstance = devLauncherController
          UpdatesControllerRegistry.controller = WeakReference(devLauncherController) // <----- HERE
        } else {
          singletonInstance = DisabledUpdatesController(context, null)
        }
        ...
}
```

This code is called after we initialise the `DevLauncherController` and initialize the instance's `updatesInterface` member:

```kt
fun createApplicationLifecycleListeners(context: Context?): List<ApplicationLifecycleListener> =
    listOf(
      object : ApplicationLifecycleListener {
        override fun onCreate(application: Application?) {
          check(application is ReactApplication)
          DevLauncherController.initialize(application)
          UpdatesControllerRegistry.controller?.get()?.let {
            DevLauncherController.instance.updatesInterface = it  // <----- HERE
            it.updatesInterfaceCallbacks = WeakReference(DevLauncherController.instance)
          }
        }
      }
    )
```

For this to work the instance of the `DevLauncherController` has to be set - which is done as a sideeffect in `DevLauncherController.initialize(application)`.

*The problem is that this side effect was not called on the old architecture.*

The side effect is called from within the `ReactHostWrapper.kt` when calling the `reactHostProvider()` function argument on initialisation. When using the old architecture this was not done as a solution to a problem with multiple reloads that was addressed in #32532.

This commit proposes a solution to this where the `DevLauncherController` making sure the necessary side effects is run at the correct time:

```kt
@JvmStatic
fun initialize(reactApplication: ReactApplication, additionalPackages: List<ReactPackage>? = null, launcherClass: Class<*>? = null) {
	...
	val reactHost = reactApplication.reactHost
	initialize(reactApplication as Context, ReactHostWrapper(reactApplication.reactNativeHost) { reactHost })
	...
}
```

Here we're calling the `reactHost` property *only once* and then pass a new lambda to the ReactHostWrapper so it can initialise without having to consider these side effects.

chrfalch committed:
Update packages/expo-dev-launcher/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

Kudo commented on packages/expo-dev-launcher/android/src/debug/java/expo/modules/devlauncher/DevLauncherController.kt:
```diff
@@ -435,7 +435,11 @@ class DevLauncherController private constructor() :
+      // Call reactHost on application to ensure we initialise it correctly and only once before we pass it to the
+      // ReactHostWrapper. Calling reactApplication.reactHost has side effects that should be run independently of
+      // whether we're on the old or new architecture.
+      val reactHost = reactApplication.reactHost
```
it doesn't look like a correct solution to me. we intentionally prevent to call `reactApplication.reactHost` on old arch mode, which will create bridgeless react instance under the hood. it would causes two react instances (both bridge and bridgeless), which is not ideal

let me figure out the root cause when i get a chance. thanks again for looking into the issue

chrfalch replied:
What it looked like to me was that the dev-packages had so many dependencies on the initialization of the reactHost (not the running app itself) - so this was created to resolve those expectations. I Agree that this seems suboptimal - but it looked a bit complex to refactor the dev-package setup to avoid it being dependent on this being done. But as usual you have a good point - Let's see if we can find a solution together.

Kudo replied:
the root cause is about initialization sequence as you shared.
```
-> read `UpdatesControllerRegistry.controller` in DevLauncherPackageDelegate (right after application onCreate)
-> set `UpdatesControllerRegistry.controller` in UpdatesController.initializeWithoutStarting() (before react instance creation)
```
of course it gets a null `UpdatesControllerRegistry.controller`

i'm proposing the fix to read `UpdatesControllerRegistry.controller` after it has been set. we can use the `onWillCreateReactInstance` point. expo-updates will ensure be executed before expo-dev-launcher in our design.
this is my patch fwiw

```patch
--- a/android/src/debug/java/expo/modules/devlauncher/DevLauncherPackageDelegate.kt
+++ b/android/src/debug/java/expo/modules/devlauncher/DevLauncherPackageDelegate.kt
@@ -11,15 +11,13 @@ import com.facebook.react.bridge.ReactApplicationContext
 import expo.modules.core.interfaces.ApplicationLifecycleListener
 import expo.modules.core.interfaces.ReactActivityHandler
 import expo.modules.core.interfaces.ReactActivityLifecycleListener
+import expo.modules.core.interfaces.ReactNativeHostHandler
 import expo.modules.devlauncher.launcher.DevLauncherReactActivityDelegateSupplier
-import expo.modules.devlauncher.modules.DevLauncherInternalModule
-import expo.modules.devlauncher.modules.DevLauncherModule
 import expo.modules.devlauncher.modules.DevLauncherAuth
-import expo.modules.core.interfaces.ReactNativeHostHandler
 import expo.modules.devlauncher.modules.DevLauncherDevMenuExtension
+import expo.modules.devlauncher.modules.DevLauncherInternalModule
+import expo.modules.devlauncher.modules.DevLauncherModule
 import expo.modules.devlauncher.react.DevLauncherReactNativeHostHandler
-import expo.modules.updatesinterface.UpdatesControllerRegistry
-import java.lang.ref.WeakReference
 
 object DevLauncherPackageDelegate {
   fun createNativeModules(reactContext: ReactApplicationContext): List<NativeModule> =
@@ -36,10 +34,6 @@ object DevLauncherPackageDelegate {
         override fun onCreate(application: Application?) {
           check(application is ReactApplication)
           DevLauncherController.initialize(application)
-          UpdatesControllerRegistry.controller?.get()?.let {
-            DevLauncherController.instance.updatesInterface = it
-            it.updatesInterfaceCallbacks = WeakReference(DevLauncherController.instance)
-          }
         }
       }
     )
--- a/android/src/debug/java/expo/modules/devlauncher/react/DevLauncherReactNativeHostHandler.kt
+++ b/android/src/debug/java/expo/modules/devlauncher/react/DevLauncherReactNativeHostHandler.kt
@@ -9,6 +9,10 @@ import com.facebook.react.modules.systeminfo.AndroidInfoHelpers
 import com.facebook.soloader.SoLoader
 import expo.modules.core.interfaces.ReactNativeHostHandler
 import expo.modules.devlauncher.DevLauncherController
+import expo.modules.updatesinterface.UpdatesControllerRegistry
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
 import java.lang.ref.WeakReference
 
 class DevLauncherReactNativeHostHandler(context: Context) : ReactNativeHostHandler {
@@ -36,4 +40,18 @@ class DevLauncherReactNativeHostHandler(context: Context) : ReactNativeHostHandl
     }
     return HermesExecutorFactory()
   }
+
+  override fun onWillCreateReactInstance(useDeveloperSupport: Boolean) {
+    super.onWillCreateReactInstance(useDeveloperSupport)
+    // On New Architecture mode, `onWillCreateReactInstance()` would be called
+    // inside `DevLauncherController.initialize()`.
+    // The `DevLauncherController.instance` is not available at this point.
+    // Posting the updates interface setup to next run loop.
+    CoroutineScope(Dispatchers.Main).launch {
+      UpdatesControllerRegistry.controller?.get()?.let {
+        DevLauncherController.instance.updatesInterface = it
+        it.updatesInterfaceCallbacks = WeakReference(DevLauncherController.instance)
+      }
+    }
+  }
 }
```

chrfalch replied:
This looks like a very nice solution - thanks for taking the time to review - I will update it and retest!

chrfalch replied:
Hi Kudo! I've tested the above patch to see if it solves the problem, but I'm not able to show the `runtimeVersion` on neither new nor old arch now. I'm seeing some strange issues with the dependency injection here where `DevLauncherController.instance.updatesInterface` is null after assigning it inside the Coroutine scope - the setter of the updatesInterface is using some kind of dependency injection that I'm finding a bit hard to debug.

I'm expecting `DevLauncherInternalModule.kt#226` to contain a valid runtimeVersion after execution.

Maybe we could sit down and see if we can look at this together when you're available?

Kudo replied:
oh i missed one call flow where launching to the app directly. please try applying this change https://github.com/expo/expo/pull/34536/commits/f9c7619de8c77e922be3a5a4bf2ccc2d79023803. hopefully it works this time

chrfalch replied:
I applied the suggested tests from you @Kudo - and it works really well.

Did some tests using Bare Expo, and everything seems to be working with the above changes:

**Before fix:** (using `main`)
- ✅ Ran `BareExpo` on new architecture → field runtimeVersion is displayed in Home/Settings in dev-client
- 🔴 Ran `BareExpo` on old architecture → field runtimeVersion fails to display in Home/Settings in dev-client

**After fix:**
- ✅ Ran `BareExpo` on new architecture → field runtimeVersion is displayed in Home/Settings in dev-client
- ✅ Ran `BareExpo` on old architecture → field runtimeVersion is displayed in Home/Settings in dev-client

Image from failing **before**:

![image](https://github.com/user-attachments/assets/b7ffa04c-869c-4e1c-a9c9-b1defa50b5fb)

Image from **after**:

![image](https://github.com/user-attachments/assets/b78a6d8b-613a-4b8e-bb9f-de1159f61133)

chrfalch committed:
codereview: applied changes after discussion with Kudo

- We shouldn't force creating a bridged reactHost - it is not necessary
- We'll update the DevLauncherController when creating the react instance
- Reset the DI Koin container when loading an app
- Ran spotless and commited changes.

Kudo commented on packages/expo-dev-launcher/android/src/debug/java/expo/modules/devlauncher/DevLauncherController.kt:
```diff
@@ -148,7 +148,7 @@ class DevLauncherController private constructor() :
       val appIntent = createAppIntent()
 
-      internalUpdatesInterface?.reset()
+      DevLauncherKoinContext.app.koin.getOrNull<UpdatesInterface>()?.reset()
```
@lukmccall since we changed the initialization order, the `DevLauncherController.instance.updatesInterface = it` will be called after this function. the default koin injection for `internalUpdatesInterface` would make it as null even later `DevLauncherController.instance.updatesInterface = it` is called.

that's the main reason to change it as `DevLauncherKoinContext.app.koin.getOrNull<UpdatesInterface>()?.reset()` here.
not sure whether the change makes sense to you?

the other question, do you know whether we still use `internalUpdatesInterface` in our code? it looks like we don't use it in [any other places.](https://github.com/search?q=repo%3Aexpo%2Fexpo%20internalUpdatesInterface&type=code) maybe we can also remove `internalUpdatesInterface` without using DI?

Kudo replied:
yep let's keep it as-is without removing `internalUpdatesInterface`. we can do it when it's necessary for a refactoring.

Kudo reviewed (approved):
approving but please wait @lukmccall for the final call
